### ISSUE TITLE
[Bug]: OpenGL setNativeSharedContext with attach() randomly not working

### ISSUE THREAD

**Issue by benkuper** [juce-framework/JUCE#1452]
### Detailed steps on how to reproduce the bug

Hello,
On the latest Juce 8 dev branch as of writing this, I’m facing an issue with sharing OpenGL contexts, tried to go deep in debug but didn’t find anything relevant…

I created a minimal program for the sake of the issue.
I have 2 components :

Component A (the main window with the titlebar) :

has OpenGLContext A and a FrameBuffer
draws a triangle into the framebuffer
draws the framebuffer twice in the window, bottom left and top right (so far so good)
Component B (the small square window on the top left)

has OpenGLContext B that has setNativeSharedContext(ContextA)
draw context A’s framebuffer twice, bottom left and top right
When I launch the program, I have 90% chance of seeing the following picture, showing white rectangles instead of the framebuffer’s texture

![Screenshot 2024-10-20 115137](https://github.com/user-attachments/assets/be8a4ff2-10c4-4d6b-8377-89edf2a55059)

To test faster, I added a keypress shortcut to do :

detach();
attach(*this);
on the second window.
I then hit the shortcut and will see that sometimes, (I’d say around 20%, but actually really random) the texture will be actually showing like the picture above. And most often going back to white rectangle.

![Screenshot 2024-10-20 115225](https://github.com/user-attachments/assets/601eee22-b17f-4287-88a0-ee30484a23f6)

I tried breaking and inspecting potential differences and changing variables inside the attach() function when it works and doesn’t work, but couldn’t find anything that was clearly different in the 2 scenarios. Doesn’t mean there aren’t, just that they’re not trivial.

I have a suspicion that the randomness of this problem is due to timing and context availability.

Interesting this to note, I tried putting CriticalSection locks around both the renderGL in the mainwindow and the attach function in the second window, and this way I never get a good texture, it will always stay white.

Any help appreciated, I’ve spent a good amount of time on this already and feel a bit lost on what I can do next :slight_smile:

I can provide the source and executable of this example if anyone wants to test, this has only been tested on Win11 x64 for now.

**EDIT :
It seems that GL is more broken than I though, even an empty project with a glGetError will throw INVALID_OPERATION, and glGetError is not clearing the flags, so  OpenGLHelpers::resetErrorState() will actually just stay forever in its while loop. I'll post another issue for this as this is a way more critical and simple issue
**

Thank you

### What is the expected behaviour?

I would really be happy with this function always working :)

### Operating systems

Windows

### What versions of the operating systems?

11

### Architectures

64-bit

### Testing on the `develop` branch

The bug is present on the `develop` branch

### Code of Conduct

- [X] I agree to follow the Code of Conduct

**Comment by benkuper** [comment]
Following the research, it seems that 
setNativeSharedContext needs to happen inside the GL thread of the shared context, otherwise there will be some randomness.
It definitely doesn't fix the problem 100%, as I managed to get my multiple external context to work with some dark timer magic, delaying calls between the registrations with this function from hell :

```
void GlContextHolder::registerSharedRenderer(OpenGLSharedRenderer* r, int delayBeforeAttach)
{
	//GenericScopedLock lock(renderLock);

	r->context.detach();

	r->context.setSwapInterval(0);
	r->context.setRenderer(r);

	context.executeOnGLThread([this, r](OpenGLContext& callerContext) {
		r->context.setNativeSharedContext(context.getRawContext());
		}, true);


	if (delayBeforeAttach == 0)
	{
		r->context.attachTo(*r->component);
		sharedRenderers.add(r);
	}
	else
	{
		Timer::callAfterDelay(delayBeforeAttach, [this, r]() {
			r->context.attachTo(*r->component);
			sharedRenderers.add(r);
			});
	}
}
```

With this function, adding what different delays to different components sharing the main context, I kinda get something almost stable, but I really don't like it..

I really wish there was a way to either detect if the context sharing has been successful, or if the access of a framebuffer is available or not from a shared context. at least I couldn't know and try to reload automatically.

If anyone has any knowledge of this, please let me know !

**Comment by reuk** [comment]
> I can provide the source and executable of this example if anyone wants to test, this has only been tested on Win11 x64 for now.

Yes, please share the example program. That will help greatly with debugging the issue.

**Comment by benkuper** [comment]
Here you go :
https://github.com/benkuper/JuceSandbox/tree/OpenGLSharedContext
branch OpenGLSharedContext

I couldn't even get it working again with executeOnGLThread..

Also, once the second window is working, it seems that adding other ones is not necessarily working the same way, as well as adding other shared context in the main window.

**Comment by benkuper** [comment]
The real program is this one : 
https://github.com/Golden-Geek/MapGyver if you want to test out

**Comment by reuk** [comment]
Thanks for the example program.

Just to check - did this ever work for you, and if so, on which JUCE version?

**Comment by benkuper** [comment]
I must say, this development is quite new and I'm not at all an OpenGL master. We started last year and had seen some problems already with Juce 7 I think to recall, but it was only few days of dev, and I started again in september.
The example software is clean enough to test on different versions quite easily I'd say, if you'd like me to test and report, let me know.

**Comment by reuk** [comment]
Thanks, I was wondering whether this used to work, because it would be easier to determine the cause of the problem if I could see the code change that introduced a regression. In this case, I tried testing with 7.0.12 and 6.1.6 but both old versions had broken behaviour.

It looks like the root of the problem is that `wglShareLists()` sometimes fails. The most common reason for this appears to be that the context is active in a different thread at the point of the wglShareLists call. So, we need to ensure that no other thread has activated the context when we make the call to wglShareLists.

This patch moves the wglShareLists call onto the GL thread. It seems to resolve the issue for me. Please could you try out this patch and see whether it helps in your situation?

[0001-OpenGL-Set-up-context-sharing-on-the-GL-thread-rathe.patch](https://github.com/user-attachments/files/17497422/0001-OpenGL-Set-up-context-sharing-on-the-GL-thread-rathe.patch)

**Comment by benkuper** [comment]
Yep that seems to be it, works all the time now ! Thank you so much, these are huge fixes :)

**Comment by reuk** [comment]
Amazing, that was a real head-scratcher. Thanks again for the demo project, it's always much easier to debug when I can see the problem first-hand.

I'll get this ready for internal review, and if all goes well, hopefully this will be released in a week or so.

**Comment by benkuper** [comment]
actually I still get flashes but not at the same moment ^^ I'll try and see when that happens, pretty sure that's not on JUCE now, more of "not-yet-drawn" framebuffer.

**Comment by reuk** [comment]
Thanks for reporting, this should be fixed on develop: https://github.com/juce-framework/JUCE/commit/4bc295241994409a77b9227943650e59d95ef3d0
